docs/react/week-3/lesson.md

```diff
@@ -444,81 +556,49 @@ function CreateAccountForm() {
 }
 ```
 
-We now have three different inputs named `username`, `email` and `password`, and we keep each entered value in state as a field in an object. The method `handleChange` is reused to keep track of changes for **all** the values. The trick here is to use the `name` of the `<input>` element to update the corresponding state.
+We now have three different inputs named `username`, `email` and `password`, and we keep each entered value in state as a field in an object.
 
-#### Updating an Object in State
+The `handleChange` function is reused to keep track of changes for **all** the values. To be able to tell which `<input>` is being updated, we check the `event.target.name` field. This corresponds to the `name` attribute on the `<input>` (e.g. `<input name="username">`).
 
-Did you spot the strange syntax in `handleChange`?
-
-We are using [object spread](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#Spread_in_object_literals) syntax, which copies properties from one object to another. We are using this to update the state object but without _mutating_ the original object.
-
-Let's break this down into 2 steps. Here we create a **new** object based on another object, and **adding** a new property ([interactive example](https://jsbin.com/suyekiwezu/edit?js,console)):
+Based on this value, we then decide which state to update:
 
 ```js
-const sherlock = {
-  name: "Sherlock Holmes",
-};
-
-const sherlockAfterMoving = {
-  ...sherlock,
-  address: "221b Baker Street",
-};
-
-console.log(sherlock);
-console.log(sherlockAfterMoving);
+if (event.target.name === "username") {
```

Comment:
Maybe this lesson is already long enough, but this could be a really nice opportunity to show a contrasting implementation where we use arrow functions as event handlers, and to remind the students once again about needing function references. Something like:

```jsx
<input onClick={event => setUsername(event.target.value)} />
<input onClick={event => setEmail(event.target.value)} />
<input onClick={event => setPassword(event.target.value)} />
```

This is something I've seen them try to do, and get stuck on a lot. Very frequently they try to write;
```jsx
<button onClick={incrementCounter()} />
```
and get confused when it doesn't work.

The fact that the single event handler function is a bit involved and fiddly, and these little arrow functions are concise, leads to a nice contrasting (and then we could maybe say something like "But if you're doing some validation or something, maybe the single function is a better approach because you can centralise logic"). Contrasting approaches, and thinking about why one is better than the other, isn't something we do much of, and would be great to do more of :)

Reply:
> Very frequently they try to write;

That is why we have such an extensive event handling section in week 2 🙂 (IME most React tutorials hardly touch on this).

But I do agree that it's one of the most common problems for our students.

> single event handler function is a bit involved and fiddly

I think the feeling was that this is a common enough pattern that it would be worth showing.

> little arrow functions are concise

We've gone back and forth on multiple different implementations in this section. I am still unsure what the best option is.

In the past, my hesitation with inline event handlers is that they are somewhat tricky to debug:

```js
<input onClick={event => console.log('foo') setPassword(event.target.value)} />
```

I also think it leads to a bad habit where students dump all the logic inline, leading to poor readability.

I however could see an argument for showing this implementation with separate function definitions first, and mentioning that they could be inlined.

> Maybe this lesson is already long enough

Yes indeed, my perception is that this lesson will already be quite challenging to fit into one day.

My proposal:

1. We switch the first implementation to multiple separate function definitions
2. We then mention that they could be inlined, as with your example
3. Finally, we show a centralised implementation (i.e. the existing one), and mention validation

I think this matches the order of importance of the concepts:

1. Handling multiple form fields
2. Practicing event handling
3. Writing form code concisely
4. Showing patterns for centralised code

Which allows the teacher to cut off if out of time.

Thoughts?

Reply:
Sounds great! :)